File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxPathSum(TreeNode root) {
        if(root == null) {
            return 0;
        }
        int[] arr = {Integer.MIN_VALUE};
        helper(root, arr);
        
        return arr[0];
    }
    
    int helper(TreeNode root, int[] arr) {
        if(root == null) {
            return 0;
        }
        
        if(root.left == null && root.right == null) {
            arr[0] = Math.max(root.val, arr[0]);
            return root.val;
        }
        int lans = helper(root.left, arr);
        int rans = helper(root.right, arr);
        
        int lsum = root.val + lans;
        int rsum = root.val + rans;
        int tsum = lans + rans + root.val;
        
        int temp = Math.max(tsum, Math.max(root.val, Math.max(lsum, rsum)));
        arr[0] = Math.max(arr[0], temp);
        return Math.max(root.val, Math.max(lsum, rsum));
        
        
    }
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int helper(TreeNode* root, int &ans) {
        if(root == NULL) {
            return 0;
        }
        // get answers from left and right
        // max with 0 is for filtering out -ve sums
        int leftAns = max(0, helper(root->left, ans));
        int rightAns = max(0, helper(root->right, ans));

        // update ans before returning a path
        ans = max(ans, leftAns + rightAns + root->val);

        // return the path (either left or right) which gives
        // the maximum sum 
        return root->val + max(leftAns, rightAns);

    }
    int maxPathSum(TreeNode* root) {
        int ans = -1e9; // ans will store the max path sum
        helper(root, ans);
        return ans;
    }
};